rat/ops.py

```diff
@@ -108,6 +240,35 @@ def scalar(self):
     def get_key(self):
         return self.name
 
+    def fold(self):
+        self.lower = self.lower.fold()
+        self.upper = self.upper.fold()
+        if self.subscript:
+            self.subscript = self.subscript.fold()
+        if not isinstance(self.lower, (RealConstant, IntegerConstant)):
+            raise ConstantFoldError(
+                f"Lower bound value must fold into a Numeric constant at compile time, but folded expression {self.lower} is not a constant!",
+                self.line_index,
+                self.column_index,
+            )
+        if not isinstance(self.upper, (RealConstant, IntegerConstant)):
+            raise ConstantFoldError(
+                f"Upper bound value must fold into a Numeric constant at compile time, but folded expression {self.upper} is not a constant!",
```

Comment:
Oh I see so the constraint checks move now that the parser knows types!

Reply:
Well I guess this isn't the constraint itself, but the checking that it is indeed a constant thingy

Reply:
Yup. we can just call `fold()` and check.

Reply:
The drawback for this is perhaps when the compiler needs to throw an error - the column index can be a bit messed up